### PR TITLE
more for fix #157: update detect vs2017 scripts to v1.11.0

### LINKED ISSUE #157: Support VS2017
This patch https://github.com/boostorg/build/commit/ae5e63a1315493dfb2fc60c4d537674fa637f898 doesn't work with VS2017RC1 

See also [Finding the location of a VC++2017 install](https://blogs.msdn.microsoft.com/vcblog/2016/10/07/compiler-tools-layout-in-visual-studio-15/)

BTW toolset should be vc141 for VS2017 and %VSxxxCOMNTOOLS% environment variable doesn't exist

### PR DESCRIPTION
**Pull request by refack**

### CODE REVIEW

**Comment by KindDragon** [comment]
Hi. I reworked your script in PR #173 to support checking `Microsoft.VisualCpp.Tools.Core` version and return only string instead of JSON

File: src/tools/vc141helper/GetKey.ps1 [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+# version: 1.11.0
+
+param (
+    [Parameter(Mandatory=$false)]
+    [string]$filter = "IsVcCompatible",
+    [Parameter(Mandatory=$false)]
+    [string]$key = "VisualCppToolsCL"
+)
+Add-Type -Path GetVS2017Configuration.cs;
+$inst = ([VisualStudioConfiguration.ComSurrogate]::QueryEx($filter))[0]
```
Comment (KindDragon) [inline src/tools/vc141helper/GetKey.ps1]: It can be empty list

Comment (refack) [inline src/tools/vc141helper/GetKey.ps1]: Changed

**Comment by refack** [comment]
@KindDragon I also refactored the JSON code to a different file, added "IsVcCompatible" predicate, that checks for the minimal VC triad: `"SDK" && "MSBuild" && "VisualCppTools"`, and streamlined the `.ps1`

**Comment by refack** [comment]
@KindDragon P.S. how did you get github to acknowledge AppYevor?

**Comment by KindDragon** [comment]
> @KindDragon I also refactored the JSON code to a different file, added "IsVcCompatible" predicate, that checks for the minimal VC triad: "SDK" && "MSBuild" && "VisualCppTools", and streamlined the .ps1

But you not checking visual studio version

> @KindDragon P.S. how did you get github to acknowledge AppYevor?

AppVeyor build every PR for this repository https://ci.appveyor.com/project/boostorg/build/build/1.0.106

BTW because of PowerShell test run 33 minutes for VS2017 and only 6 min for other versions. I will try switching to `vswhere` to speed up tests

**Comment by refack** [comment]
> But you not checking visual studio version

What do you mean? You meen future proofing (since for now it's only 15.0)?

> BTW because of PowerShell tests run 33 minutes for VS2017 and only 6 min for other versions. I will try switching to vswhere to speed up tests

Let me look into that... (since the actual `bootstrap.bat` and `bjam` work reasonably fast...)

**Comment by KindDragon** [comment]
> What do you mean? You meen future proofing (since for now it's only 15.0)?

It will probably work for next VS after 2017

> Let me look into that... (since the actual bootstrap.bat and bjam work reasonably fast...)

I see in Process Explorer spawning a lot of PowerShell executions. I think each test run your script many times

**Comment by refack** [comment]
> I see in Process Explorer spawning a lot of PowerShell executions. I think each test run your script many times

That's a bug i've seen in `bootstrap.bat` when something fails...

**Comment by refack** [comment]
I.E. look at how it fails https://ci.appveyor.com/project/boostorg/build/build/1.0.106/job/1fabtlet1fdlmr6o
`******  B A T C H   R E C U R S I O N  exceeds STACK limits ******`

**Comment by KindDragon** [comment]
> I.E. look at how it fails https://ci.appveyor.com/project/boostorg/build/build/1.0.106/job/1fabtlet1fdlmr6o
> ****** B A T C H R E C U R S I O N exceeds STACK limits ******

I fixed this in my PR by splitting build.bat to several files. I tried many options, but only it helped

**Comment by refack** [comment]
Yea I can relate. `cmd` is super fickle 🤦

**Comment by refack** [comment]
Green on [my fork](https://ci.appveyor.com/project/refack/build/build/1.0.8) in good time

@grafikrobot you can cancel all others builds

**Comment by refack** [comment]
@grafikrobot would you consider adding the COM helper tools as a [git submodule](https://github.com/node4good/msvs-com-helper)?

File: test/test_all.py [hunk 2]
```diff
@@ -284,6 +283,7 @@ def reorder_tests(tests, first_test):
 
 if os.name == "posix":
     tests.append("symlink")
+    tests.append("link")
```
Comment (KindDragon) [inline test/test_all.py]: This test works on Windows for me, but not on AppVeyor

File: src/engine/guess_toolset.bat [hunk 3]
```diff
@@ -32,11 +33,24 @@ if NOT "_%VS150COMNTOOLS%_" == "__" (
...
+SET vsCOMhelper_cmd="%~dp0..\tools\vsCOMhelper\get_key_helper.cmd" IsVcCompatible
+for /f "tokens=*" %%A in ('cmd /D /S /C "%vsCOMhelper_cmd% InstallationPath"') do call :vsCOMhelper_2 "%%A"
+IF NOT DEFINED BOOST_JAM_TOOLSET goto :After_vsCOMhelper
+IF NOT DEFINED BOOST_JAM_TOOLSET_ROOT goto :After_vsCOMhelper
+goto :eof
+
+:vsCOMhelper_2
+IF "_%~1_"=="__" goto :eof
+set "BOOST_JAM_TOOLSET_ROOT=%~1\VC\"
+for /f "tokens=*" %%A in ('cmd /D /S /C "%vsCOMhelper_cmd% VCToolsVersionCode"') do if NOT "_%%A_"=="__" (
```
Comment (KindDragon) [inline src/engine/guess_toolset.bat]: Who will check version from `VCToolsVersionCode`?

File: src/engine/config_toolset.bat [hunk 4]
```diff
@@ -165,21 +166,25 @@ if "_%BOOST_JAM_TOOLSET_ROOT%_" == "__" (
     if NOT "_%VS150COMNTOOLS%_" == "__" (
         set "BOOST_JAM_TOOLSET_ROOT=%VS150COMNTOOLS%..\..\VC\"
     ))
-SET cl_path_cmd="%~dp0..\tools\vc141helper\cl_path.cmd"
+
+IF DEFINED EXPENSIVE_POWERSHELL2 goto :Skip_VC1410
+SET EXPENSIVE_POWERSHELL2=1
+SET vsCOMhelper_cmd="%~dp0..\tools\vsCOMhelper\get_key_helper.cmd" "IsVcCompatible,VisualCppToolsVersionMinor=14.10" "InstallationPath"
```
Comment (KindDragon) [inline src/engine/config_toolset.bat]: I can't find where Powershell/CSharp script check this version

File: src/engine/guess_toolset.bat (hunk 3 above)
Comment (refack) [inline src/engine/guess_toolset.bat]: So it's forward compatible. Bootstrap will fail with `Unknown toolset: %BOOST_JAM_TOOLSET%`

File: src/engine/config_toolset.bat (hunk 4 above)
Comment (refack) [inline src/engine/config_toolset.bat]: https://github.com/boostorg/build/pull/179/files#diff-88a445c97b267cf7aa36e1880e333390R17

File: test/test_all.py (hunk 2 above)
Comment (refack) [inline test/test_all.py]: Fails for me locally as well. I opened #185, will look into it soon.

File: src/engine/config_toolset.bat (hunk 4 above)
Comment (refack) [inline src/engine/config_toolset.bat]: My concept is; `C#` only does marshaling of data from COM, `powershell` does filtering / projecting.

File: src/engine/guess_toolset.bat [hunk 5]
```diff
@@ -1,4 +1,5 @@
-@ECHO OFF
+@IF NOT DEFINED DEBUG_BATCH @ECHO OFF
```
Comment (KindDragon) [inline src/engine/guess_toolset.bat]: I just comment `@echo off` if I need debugging 😄

Comment (refack) [inline src/engine/guess_toolset.bat]: It goes well with https://github.com/boostorg/build/pull/179/files#diff-180360612c6b8c4ed830919bbb4dd459R14

File: src/engine/guess_toolset.bat (hunk 3 above)
Comment (KindDragon) [inline src/engine/guess_toolset.bat]: I just can't find where get_key_helper.cmd checking `VCToolsVersionCode`

File: src/engine/guess_toolset.bat (hunk 5 above)
Comment (refack) [inline src/engine/guess_toolset.bat]: BTW: stole the concept from VS2017 `VsDevCmd.bat` line 3: `@if NOT "%VSCMD_DEBUG%" GEQ "3" @echo off`

File: src/engine/config_toolset.bat (hunk 4 above)
Comment (refack) [inline src/engine/config_toolset.bat]: @KindDragon I took the filtering idea from you, it's a very good idea 👍 I'm using it in all other projects...

File: appveyor.yml [hunk 6]
```diff
@@ -11,27 +11,30 @@ image:
...
-    & python test_all.py msvc $env:TEST_ALL_EXTRAS --preserve
-    $TestExitCode = $LastExitCode
-    if (Test-Path failed_test)
-    {
-        & 7z a failed_test.zip failed_test\*
-        Push-AppveyorArtifact failed_test.zip
-    }
-    if ($TestExitCode -ne 0) { $host.SetShouldExit($TestExitCode) }
+    
+    python test_all.py msvc $env:TEST_ALL_EXTRAS --preserve
```
Comment (KindDragon) [inline appveyor.yml]: It should begin with &

File: src/engine/guess_toolset.bat (hunk 3 above)
Comment (refack) [inline src/engine/guess_toolset.bat]: The full command will be `"%~dp0..\tools\vsCOMhelper\get_key_helper.cmd" IsVcCompatible VCToolsVersionCode`
first arg `IsVcCompatible` is [a filter ](https://github.com/boostorg/build/pull/179/files#diff-88a445c97b267cf7aa36e1880e333390R15)
second arg `VCToolsVersionCode` is[ a projection](https://github.com/boostorg/build/pull/179/files#diff-88a445c97b267cf7aa36e1880e333390R27)

File: appveyor.yml (hunk 6 above)
Comment (refack) [inline appveyor.yml]: Not sure why? This works (as it should).

P.S. __&__ operator does not parse variables.
From the docs:

__&__ Call operator
==
Run a command, script, or script block.
The call operator, also known as the "invocation operator," lets you run commands that are stored in variables and represented by strings.
__Because the call operator does not parse the command, it cannot interpret command parameters.__

```
C:\PS> $c = "get-executionpolicy"
C:\PS> $c
get-executionpolicy
C:\PS> & $c
AllSigned
```

Comment (KindDragon) [inline appveyor.yml]: You right

Comment (refack) [inline appveyor.yml]: http://stackoverflow.com/questions/42382951/whats-the-command-of-call-operator
Learned something new: You need the __&__ if you want to run a command that's not in enf:Path with a space in it's name, i.e.:
```
& "C:\bin\dev\VS\SDKs\10\bin\x64\rc.exe" $param1 $param2 $param3
```

Comment (refack) [inline appveyor.yml]: I already know you can use it to run `{ script blocks }`
```
$reusable = { msbuild "node.sln" /verylong:parameter1 /verylong:parameter2 } //This one doesn't run
& $reusable 'nobuild';
& $reusable 'build';
& $reusable 'test';

**Comment by refack** [comment]
Made obsolete by official install of `vswhere`

**Comment by AndrewPardoe** [comment]
The officially supported vswhere tool is now available as part of the install starting today with Visual Studio 15.2 preview 2: https://blogs.msdn.microsoft.com/heaths/2017/04/21/vswhere-is-now-installed-with-visual-studio-2017/. This doesn't fix all the existing installs of VS 2017 but going forward you can now detect the VS install locations and whether C++ tools are available from a scripted environment. I hope this helps your scenario.

**Pull request closed without merging**